### PR TITLE
Heat pump: Tests, Translation, Docstrings, Relocation

### PR DESCRIPTION
**Pull request by fortrieb**
Heat pump class is refactored for this PR. Following things where done:

- Translation to english
- Adding type hints to methods
- Check inputs for methods
- Adding Python doc strings to methods
- Add more units tests to cover all methods
- Remove `class` prefix of python file

### CODE REVIEW

**Comment by michaelosthege** [comment]
Run these to fix code style:
* `pip install -r requirements-dev.txt`
* `pre-commit run --all`

File: modules/heatpump.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
+import logging
+from typing import List
+
+
+class HeatPump:
+    MAX_HEATOUTPUT = 5000  # Maximum heating power in watts
+    BASE_HEATPOWER = 235.0  # Base heating power value
+    TEMPERATURE_COEFFICIENT = -11.645  # Coefficient for temperature
+    COP_BASE = 3.0  # Base COP value
+    COP_COEFFICIENT = 0.1  # COP increase per degree
```
Comment (michaelosthege) [inline modules/heatpump.py:10]: This is the docstring syntax for class attributes:

```suggestion
    MAX_HEATOUTPUT = 5000
    """Maximum heating power in watts"""

    BASE_HEATPOWER = 235.0
    """Base heating power value"""

    TEMPERATURE_COEFFICIENT = -11.645
    """Coefficient for temperature"""

    COP_BASE = 3.0
    """Base COP value"""

    COP_COEFFICIENT = 0.1
    """COP increase per degree"""
```

File: requirements-dev.txt [hunk 2]
```diff
@@ -1,2 +1,2 @@
 pytest==8.3.3
-pre-commit
+pre-commit==3.8.0
```
Comment (michaelosthege) [inline requirements-dev.txt:2]: For some dependencies the exact versions don't matter and pinning them can even increase maintenance workload:
* Dependabot making PRs about every small version increment that doesn't actually matter
* Devs won't actually `pip install -r requirements-dev.txt` every time they continue working on the project

For the `requirements.txt` I would even consider it bad practise to pin all versions exactly, because it makes it very hard to install the library into an environment also containing other packages.

Fixing it for the Docker-build, however, would be good practice (as long as it's kept updated by dependabot).

File: tests/conftest.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import pytest
+
+from modules.heatpump import HeatPump
+
+
+@pytest.fixture(scope="function")
+def hp_5kw_24h() -> HeatPump:
+    """Heat pump with 5 kw heating output and 24 h prediction"""
+    return HeatPump(5000, 24)
```
Comment (michaelosthege) [inline tests/conftest.py:9]: why not put it in `test_heatpump` to keep the related things together?

File: tests/test_heatpump.py [hunk 4]
```diff
@@ -1,17 +1,44 @@
...
-@pytest.fixture(scope="function")
-def heatpump() -> Heatpump:
-    """Heatpump with 5 kw heating power and 24 h prediction"""
-    return Heatpump(5000, 24)
+class TestHeatPump:
+    def test_cop(self, hp_5kw_24h: HeatPump):
+        """Testing calculate COP for variouse outside temperatures"""
+        assert hp_5kw_24h.calculate_cop(-10) == 2.0, "COP for -10 degree isn't correct"
+        assert hp_5kw_24h.calculate_cop(0) == 3.0, "COP for 0 degree isn't correct"
+        assert hp_5kw_24h.calculate_cop(10) == 4.0, "COP for 10 degree isn't correct"
```
Comment (michaelosthege) [inline tests/test_heatpump.py:11]: the messages are trivial, given that the assert statement is already quite self-explanatory
```suggestion
        assert hp_5kw_24h.calculate_cop(-10) == 2.0
        assert hp_5kw_24h.calculate_cop(0) == 3.0
        assert hp_5kw_24h.calculate_cop(10) == 4.0
```

File: tests/test_heatpump.py [hunk 5]
```diff
@@ -1,17 +1,44 @@
...
+        # Check edge case for outside temperature
+        out_temp_min = -100.1
+        out_temp_max = 100.1
+        with pytest.raises(ValueError) as err:
+            hp_5kw_24h.calculate_cop(out_temp_min)
+        assert str(out_temp_min) in str(err.value)
+        with pytest.raises(ValueError) as err:
+            hp_5kw_24h.calculate_cop(out_temp_max)
+        assert str(out_temp_max) in str(err.value)
```
Comment (michaelosthege) [inline tests/test_heatpump.py:20]:
```suggestion
        with pytest.raises(ValueError, match=f"'{out_temp_min}' not in range"):
            hp_5kw_24h.calculate_cop(out_temp_min)
        with pytest.raises(ValueError, match=f"'{out_temp_max}' not in range"):
            hp_5kw_24h.calculate_cop(out_temp_max)
```

File: tests/test_heatpump.py [hunk 6]
```diff
@@ -1,17 +1,44 @@
...
+        """Testing calculate of heating output"""
+        assert (
+            hp_5kw_24h.calculate_heating_output(-10.0) == 5000
+        ), "Wrong output at -10.0 Celsius"
+        assert (
+            hp_5kw_24h.calculate_heating_output(0.0) == 5000
+        ), "Wrong output at 0.0 Celsius"
+        assert hp_5kw_24h.calculate_heating_output(10.0) == pytest.approx(
+            4939.583
+        ), "Wrong output at 10.0 Celsius"
```
Comment (michaelosthege) [inline tests/test_heatpump.py:32]:
```suggestion
        assert hp_5kw_24h.calculate_heating_output(-10.0) == 5000
        assert hp_5kw_24h.calculate_heating_output(0.0) == 5000
        assert hp_5kw_24h.calculate_heating_output(10.0) == pytest.approx(4939.583)
```

File: tests/test_heatpump.py [hunk 7]
```diff
@@ -1,17 +1,44 @@
...
+        """Testing calculation of heating power"""
+        assert (
+            hp_5kw_24h.calculate_heat_power(-10.0) == 2104
+        ), "Wrong power at -10.0 Celsius"
+        assert (
+            hp_5kw_24h.calculate_heat_power(0.0) == 1164
+        ), "Wrong power at 0.0 Celsius"
+        assert (
+            hp_5kw_24h.calculate_heat_power(10.0) == 548
+        ), "Wrong power at 10.0 Celsius"
```
Comment (michaelosthege) [inline tests/test_heatpump.py:44]:
```suggestion
        assert hp_5kw_24h.calculate_heat_power(-10.0) == 2104
        assert hp_5kw_24h.calculate_heat_power(0.0) == 1164
        assert hp_5kw_24h.calculate_heat_power(10.0) == 548
```

File: modules/heatpump.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
+import logging
+from typing import List
```
Comment (michaelosthege) [inline modules/heatpump.py:2]:
```suggestion
from typing import List, Sequence
```

File: modules/heatpump.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+                + 1.62 * outside_temperature_celsius**2.0
+            )
+        else:
+            err_msg = f"Outside temperature '{outside_temperature_celsius}' not in range (min: -100 Celsius, max: 100 Celsius) "
+            self.log.error(err_msg)
+            raise ValueError(err_msg)
+
+    def simulate_24h(self, temperatures: List[float]) -> List[float]:
+        """Simulate power data for 24 hours based on provided temperatures."""
+        power_data = List[float]
```
Comment (michaelosthege) [inline modules/heatpump.py:103]: Rule of thumb: Use invariant types (`Sequence`, `Collection`, `Iterable`, `Mapping`, ...) for parameters and covariant types (`List`, `Tuple`, `Dict`, ...) for return values.
This keeps the code flexible while improving the utility of type information.

In this example: `.simulate_24h(temperatures=(1,2,3,...,24))` works with the implementation, but wouldn't be allowed if `List[float]` was required for the input.

```suggestion
    def simulate_24h(self, temperatures: Sequence[float]) -> List[float]:
        """Simulate power data for 24 hours based on provided temperatures."""
        power_data: List[float] = []
```

**Review by michaelosthege — COMMENTED** [review]
Good refactoring!

Mid-term I see the issue of defining heat pumps with different properties, or even multiple heat pumps.
But I think this is a topic that should be covered by a discussion before making implementation changes.

File: requirements-dev.txt (hunk 2 above)
Comment (fortrieb) [reply requirements-dev.txt:2]: Thanks for the explanation.

File: tests/conftest.py (hunk 3 above)
Comment (fortrieb) [reply tests/conftest.py:9]: Thought we can place all fixture in one place. Maybe adding a subfolder `fixtures` inside `tests` which can be used for that.

**Review by michaelosthege — APPROVED** [review]
